transcom/mymove
# B19770 INT Change PPM Status

## Agility ticket
[B19770](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-19770)

## Summary

The following PPM shipment status values were to be changed:
* 'NEEDS_PAYMENT_APPROVAL' to 'NEEDS_CLOSEOUT'
* 'PAYMENT_APPROVED' to 'CLOSEOUT_COMPLETE'

Because the vars that held these values were the same name as the values, when the values change, the vars need to be updated too.

These values are also stored in the database.  Therefore migration SQL was added to change existing data to the new values and to update the enum of the column where the statuses are stored.


### How to test
Test 1:
Verify the DB change.
Run the 2 SQL files in this change in the following order:
* migrations/apps/schema/20240531152430_migrate_data_to_new_ppm_statuses.up.sql
* migrations/app/schema/20240531152526_remove_ppm_statuses_no_longer_used.up.sql
This will migrate existing data and re-declare the status column.

Note: 
If you don't mind blowing away your db, just run "make db_dev_fresh".
It will run all the migration SQL.

Run the following SQL:
```
SELECT e.enumlabel AS enum_value
  FROM pg_type t
       JOIN pg_enum e 
         ON t.oid = e.enumtypid
       JOIN pg_catalog.pg_namespace n
         ON n.oid = t.typnamespace
WHERE t.typname = 'ppm_shipment_status'
ORDER BY e.enumlabel;
``` 

Confirm the following statuses appear:
* CLOSEOUT_COMPLETE
* DRAFT
* NEEDS_ADVANCE_APPROVAL
* NEEDS_CLOSEOUT
* SUBMITTED
* WAITING_ON_CUSTOMER


Test 2:
* Customer: Create a PPM move
* Customer: Submit move
* SC:  approve move
* Customer: upload ppm documentation

Run the query: 
```
select m.locator as move_code
       , ps.status
  from moves m 
       inner join mto_shipments ms
               on m.id = ms.move_id
       left join ppm_shipments ps
               on ms.id = ps.shipment_id
       inner join orders o
               on m.orders_id = o.id
 where 1 = 1
   and m.locator = 'ABC123'
   and ms.shipment_type = 'PPM'
order by m."locator"
         , ms.id 
         , ms.shipment_type 
```
Change "ABC123" to your move code.

* Status in the db should be 'NEEDS_CLOSEOUT'
* SC: Review Docs, approve all docs, confirm
* Rerun the SQL.  Status should now be 'CLOSEOUT_COMPLETE'

## Screenshots
None.
